### ISSUE TITLE
Suggest `pub extern foo;` if any part of `foo` is in this crate's public API

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust-clippy#3553]
As an example: `kiss3d` uses a specific version of `nalgebra`. In order to use `kiss3d` one needs to have the matching version of `kiss3d` and `nalgebra` in their `Cargo.toml`. It is more idiomatic to reexport `nalgebra` from `kiss3d` so that users just need ot have `kiss3d` in their `Cargo.toml` and then access `nalgebra` via `kiss3d::nalgebra` or similar.

**Comment by matthiaskrgr** [comment]
````
cargo new a
cd a
echo "kiss3d = \"0.17.0\"" >> Cargo.toml
# kiss3d imports nalgebra as na
sed -i '1s/^/use kiss3d::na;\n/' src/main.rs
cargo check
````

warns

````
error: extern crate `na` is private, and cannot be re-exported (error E0365), consider declaring with `pub`
 --> src/main.rs:1:5
  |
1 | use kiss3d::na;
  |     ^^^^^^^^^^
  |
  = note: #[deny(pub_use_of_private_extern_crate)] on by default
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #34537 <https://github.com/rust-lang/rust/issues/34537>

````
suggesting that this is not very idiomatic...?
Or would this be fine if kiss3d used something like `pub use nalgebra as na;` ?

**Comment by matthiaskrgr** [comment]
I also think this would also add some unnecessary complexity to the crate as a maintainer not only would have to keep an eye on the crates dependencies but also in the dependencies of the dependencies and watch out of breaking things.

**Comment by oli-obk** [comment]
> Or would this be fine if kiss3d used something like pub use nalgebra as na; ?

Yes, that is what I'm suggesting. A lint that would trigger inside `kiss3d` to suggest adding `pub`

> I also think this would also add some unnecessary complexity to the crate as a maintainer not only would have to keep an eye on the crates dependencies but also in the dependencies of the dependencies and watch out of breaking things.

Right now you have this complexity, but you're pushing the work to the users of your crate. By reexporting you are saving everyone some time.

**Comment by matthiaskrgr** [comment]
>Yes, that is what I'm suggesting. A lint that would trigger inside kiss3d to suggest adding pub
Ah, I misunderstood the ticket! :+1:

**Comment by thomaseizinger** [comment]
> I also think this would also add some unnecessary complexity to the crate as a maintainer not only would have to keep an eye on the crates dependencies but also in the dependencies of the dependencies and watch out of breaking things.

Also, by using an external type in your public API, you are already accepting that breaking changes to that type are breaking changes to your public API. If you don't want that, you shouldn't use external types in your public API.

In regards to this ticket, what do people think of narrowing the scope of this lint? I was thinking, instead of suggesting to re-export the whole crate, the lint could just warn you that you _have_ an external type in your public API. Users could then choose to:

- Just re-export that particular type
- Ignore the warning
- Re-export the whole crate

**Comment by epage** [comment]
@triagebot label -T-cargo

**Comment by emilk** [comment]
This would be a GREAT advantage to the Rust eco-system.

Not only would it set a good precedence of exporting crates that are in the public API, it would also trigger library authors to sometimes think twice before they export an 3rd party crate in their public API.

The only open question is what to export, and how.

@thomaseizinger makes a good point in that it would sometimes make sense to only export _parts_ of the 3rd party crate you're using. So I suggest this approach:

```rs
// my_crate/src/lib.rs

// This alone should cause a warning
pub fn foo(position: glam::Vec3) { … }

// This should be enough to silence the warning:
pub use Vec3 = glam::Vec3;

// This should also work:
pub use glam;

// And so should this
pub mod external {
    pub use glam;
}
```

Maybe something for @GuillaumeGomez ? :)

**Comment by GuillaumeGomez** [comment]
Still catching up from all the all-hands items still in my TODO list so don't know when I'll get to it. ^^'

**Comment by epage** [comment]
> it would also trigger library authors to sometimes think twice before they export an 3rd party crate in their public API.

public/private dependencies is supposed to help with that problem.  There are currently holes in the rustc lint but when its done and stabilized, you will get a warning if you do not set `public = true` in a dependency but an item from it is mentioned in your public API.
